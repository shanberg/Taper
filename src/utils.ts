export function formatDate(date, style = 'conciseUS') {
let options;

  switch (style) {
    case 'conciseUS':
      options = { month: 'numeric', day: 'numeric' };
      break;
    case 'conciseEU':
      options = { day: 'numeric', month: 'numeric', year: '2-digit' };
      break;
    case 'longform':
      options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
      break;
    // Add more styles as needed
    default:
      options = {};
  }

  return new Intl.DateTimeFormat('default', options).format(new Date(date));
}

export const isRowValid = (row): boolean => {
  return row.dose <= 0 || row.daysForDose <= 0
}

export const isRowPlaceholder = (row): boolean => {
  return row.dose === 0 && row.daysForDose === 0
}


// Helper function to format dates as 'YYYY-MM-DD'
export function yyyymmdd(date) {
  const d = new Date(date),
        month = '' + (d.getMonth() + 1),
        day = '' + d.getDate(),
        year = d.getFullYear();

  return [year, month.padStart(2, '0'), day.padStart(2, '0')].join('-');
}