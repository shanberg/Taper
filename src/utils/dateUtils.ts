import { TaperDate } from '../TaperDate';

export const isDateThisYear = (date: ScheduleDate): boolean => {
  return date.getFullYear() === new Date().getFullYear();
};

/** Higher-order function to create a cached formatter */
function createCachedFormatter() {
  const dateTimeFormatCache = new Map();

  return function (date: ScheduleDate, lang: string = 'en-US'): LocaleDate {
    const yearFormat = isDateThisYear(date) ? undefined : 'numeric';
    const cacheKey = `${lang}-${yearFormat}`;

    // Check if the formatter is already in the cache
    if (!dateTimeFormatCache.has(cacheKey)) {
      // Create a new formatter and store it in the cache
      const formatter = new Intl.DateTimeFormat(lang, {
        month: 'short',
        day: 'numeric',
        year: yearFormat
      });
      dateTimeFormatCache.set(cacheKey, formatter);
    }

    // Use the cached formatter
    const formatter = dateTimeFormatCache.get(cacheKey);
    return formatter.format(date) as LocaleDate;
  };
}

// Create a cached version of formatDate
export const cachedFormatDate = createCachedFormatter();


export function calculateSegmentStartAndEndDates(schedule: Schedule, index: number): SegmentWithStartEndDate {
  const segment = schedule.segments[index];
  const taperStartDate = new TaperDate(schedule.startDate);
  const totalDaysForStartDate =
    schedule.segments
      .slice(0, index)
      .reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
    index;
  taperStartDate.incrementByDays(totalDaysForStartDate);
  const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
  taperEndDate.incrementByDays(segment.daysForDose - 1);

  return {
    segment,
    segmentStartDate: taperStartDate.toScheduleDate(),
    segmentEndDate: taperEndDate.toScheduleDate()
  }
}
