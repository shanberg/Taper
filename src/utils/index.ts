export * from './dateUtils';
export * from './stepUtils';
export * from './scheduleUtils';
export * from './serializationUtils';
export * from './textUtils';
export * from './languageUtils';