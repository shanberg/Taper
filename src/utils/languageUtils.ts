import { LANGUAGES } from '../consts';

export function getLanguageFromKey(languageKey: string): Language {
  return LANGUAGES.find((l) => l.lang === languageKey) || LANGUAGES[0];
}