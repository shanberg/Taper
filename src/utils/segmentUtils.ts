export const isSegmentInvalid = (segment: Segment): boolean => {
  return segment.dose <= 0 || segment.daysForDose <= 0;
};

/** Returns true if dose and daysForDose are both 0 */
export const isSegmentPlaceholder = (segment: Segment): boolean => {
  if (!segment) return false;
  return segment.dose === 0 && segment.daysForDose === 0;
};

export const sumSegmentsDose = (segments: Segment[]): number => {
  return segments.reduce((sum, segment) => sum + segment.dose * segment.daysForDose, 0);
};

export const sumSegmentsDays = (segments: Segment[]): number => {
  return (
    segments.reduce((sum, segment) => sum + segment.daysForDose, 0)
  );
};

export function isSegmentDirectlyAfterPlaceholder(segments: Segment[], index: number) {
  const segment = segments[index]
  if (!segment) return false;

  if (index === -1) {
    return false;
  }

  const prevSegment = segments[index - 1];
  if (!prevSegment) return false;
  return isSegmentPlaceholder(prevSegment);
}

export function segmentIsOrDirectlyAfterPlaceholder(segments: Segment[], index: number) {
  const segment = segments[index]

  if (!segment) return false;
  if (isSegmentPlaceholder(segment)) {
    return true;
  }
  if (isSegmentDirectlyAfterPlaceholder(segments, index)) {
    return true;
  }
  return false;
}