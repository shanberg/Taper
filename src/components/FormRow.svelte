<script lang="ts">
	import { createEventDispatcher } from 'svelte';
	import { isRowInvalid, isRowPlaceholder } from '../utils';
	export let tableData;
	export let row: Row;
	export let index: number;

	let rowStartDate: Date;
	let rowEndDate: Date;
	let rowIsPlaceholder: boolean = false;
	let rowIsOnlyRealRow: boolean;
	let invalid: boolean;

	const dispatch = createEventDispatcher();

	$: rowIsOnlyRealRow = tableData.length === 2;
	$: rowIsPlaceholder = isRowPlaceholder(row);
	$: isInvalid = !rowIsPlaceholder && isRowInvalid(row);

	function handleDoseChange(event: Event) {
		const target = event.target as HTMLInputElement;
		dispatch('change', { ...row, dose: parseFloat(target.value) });
	}

	function handleDaysForDoseChange(event: Event) {
		const target = event.target as HTMLInputElement;
		dispatch('change', { ...row, daysForDose: parseInt(target.value) });
	}

	function handleRemoveRow() {
		dispatch('removeRow', index);
	}
</script>

<tr class="row {rowIsPlaceholder ? 'placeholder' : ''} {isInvalid ? 'isInvalid' : ''}">
	<td class="dose">
		<input
			min={1}
			step={row.dose > 5 ? 1 : 0.25}
			type="number"
			inputmode="decimal"
			bind:value={row.dose}
			on:change={handleDoseChange}
		/>
	</td>
	<td class="days">
		<input
			min={1}
			type="number"
			inputmode="decimal"
			bind:value={row.daysForDose}
			on:change={handleDaysForDoseChange}
		/>
	</td>
	<td class="delete">
		<button
			disabled={rowIsOnlyRealRow}
			title="Remove this step"
			class="remove-btn"
			on:click={handleRemoveRow}>Ã—</button
		>
	</td>
</tr>

<style>
	input {
		width: 100%;
		height: 100%;
		border: none;
		outline: none;
	}

	.placeholder {
		color: var(--color-fg-muted);
		user-select: none;

		& input {
			background: transparent;
		}
	}

	.isInvalid td input {
		background: var(--color-bg-error);
	}

	.isInvalid:focus-within td input {
		background: var(--color-bg-error-muted);
	}

	tr:last-child .remove-btn {
		display: none;
	}

	.remove-btn {
		width: 1.5rem;
		height: 100%;
		border: 0;
		margin-left: 1px;
		border-radius: var(--control-radius);
		transition: all var(--control-transition-duration) ease-in-out;
		color: inherit;
		background: transparent;
		cursor: pointer;
		opacity: 0;
		padding: 0;
		text-align: center;
	}

	.remove-btn:disabled {
		cursor: not-allowed;
	}

	/* hide remove button on last row */

	tr:focus-within .remove-btn,
	tr:hover .remove-btn,
	.remove-btn:focus {
		outline: none;
		opacity: 1;
	}

	.remove-btn:hover:not(:disabled),
	.remove-btn:focus:not(:disabled) {
		color: var(--color-fg-error);
		background: var(--color-bg-error-muted);
	}
</style>
