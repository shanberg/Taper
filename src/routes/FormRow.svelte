<script>
  import { createEventDispatcher } from 'svelte';
  import { formatDate, isRowValid, isRowPlaceholder } from '../utils';
  export let tableData;
  export let row;
  export let index;

  const dispatch = createEventDispatcher();

  let rowStartDate
  let rowEndDate
  let rowIsPlaceholder = false

  $: rowIsOnlyRealRow = tableData.length === 2
  $: rowIsPlaceholder = isRowPlaceholder(row)
  $: invalid = !rowIsPlaceholder && isRowValid(row);

  function handleDoseChange(event) {
    dispatch('change', { ...row, dose: parseFloat(event.target.value) });
  }

  function handleDaysForDoseChange(event) {
    dispatch('change', { ...row, daysForDose: parseInt(event.target.value) });
  }

  function handleRemoveRow() {
    dispatch('removeRow', index);
  }
  
</script>

<tr class="row {rowIsPlaceholder ? 'placeholder' : ''} {invalid ? 'invalid' : ''}">
  <td class="dose">
    <input 
      min={1}
      step={row.dose > 5 ? 1 : 0.25}
      type="number"
      inputmode="decimal"
      bind:value={row.dose}
      on:change={handleDoseChange} 
    />
  </td>
  <td class="days">
    <input
      min={1}
      type="number" 
      inputmode="decimal"
      bind:value={row.daysForDose}
      on:change={handleDaysForDoseChange} 
    />
  </td>
  <td class="delete">
    <button disabled={rowIsOnlyRealRow ? "disabled" : undefined} title="Remove this step" class="remove-btn" on:click={handleRemoveRow}>Ã—</button>
  </td>
</tr>

<style>

  input {
    width: 100%;
    height: 100%;
    border: none;
    outline: none;
  }

  .placeholder {
    color: var(--color-fg-muted);
    user-select: none;

    & input {
      background: transparent;
    }
  }

  .invalid td input {
    background: var(--color-bg-error);
  }

  .invalid:focus-within td input {
    background: var(--color-bg-error-muted);
  }

  tr:last-child .remove-btn {
    display: none;
  }

  .remove-btn {
    width: 1.5rem;
    height: 100%;
    border: 0;
    margin-left: 1px;
    border-radius: var(--control-radius);
    transition: all var(--control-transition-duration) ease-in-out;
    color: inherit;
    background: transparent;
    cursor: pointer;
    opacity: 0;
    padding: 0;
    text-align: center;
  }

  .remove-btn:disabled {
    cursor: not-allowed;
  }

  /* hide remove button on last row */

  tr:focus-within .remove-btn,
  tr:hover .remove-btn, .remove-btn:focus {
    outline: none;
    opacity: 1;
  }

  .remove-btn:hover:not(:disabled), .remove-btn:focus:not(:disabled) {
    color: var(--color-fg-error);
    background: var(--color-bg-error-muted);
  }

</style>