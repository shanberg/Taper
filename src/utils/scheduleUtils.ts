import { TaperDate } from '../TaperDate';
import { TEMPLATES, DEFAULT_LANGUAGE_KEY, DEFAULT_TEMPLATE_KEY } from '../consts';
import { isSegmentInvalid, sumSegmentsDays, sumSegmentsDose, isSegmentPlaceholder } from './segmentUtils';
import { formatSegmentText } from './textUtils';
import { getLanguageFromKey } from './languageUtils';

const PLACEHOLDER_SEGMENT: Segment = { dose: 0, daysForDose: 0 };

export function createInitialSchedule(): Schedule {
  return {
    segments: [...TEMPLATES.Default, PLACEHOLDER_SEGMENT],
    startDate: new TaperDate().toScheduleDate(),
    templateKey: DEFAULT_TEMPLATE_KEY,
    languageKey: DEFAULT_LANGUAGE_KEY,
    displayMode: "segments"
  };
}

export function isValidSchedule(schedule: Schedule): boolean {
  if (!schedule) {
    throw new Error("No schedule provided");
  }
  return !schedule.segments.slice(0, schedule.segments.length - 1).some(s => isSegmentInvalid(s));
}

export function calculateScheduleSummary(schedule: Schedule): string {
  return `${sumSegmentsDose(schedule.segments)}mg over ${sumSegmentsDays(schedule.segments)} days`
};

export function calculateSegmentStartAndEndDates(schedule: Schedule, index: number): SegmentWithStartEndDate {
  const segment = schedule.segments[index];
  const taperStartDate = new TaperDate(schedule.startDate);
  const totalDaysForStartDate =
    schedule.segments
      .slice(0, index)
      .reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
    index;
  taperStartDate.incrementByDays(totalDaysForStartDate);
  const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
  taperEndDate.incrementByDays(segment.daysForDose - 1);

  return {
    segment,
    segmentStartDate: taperStartDate.toScheduleDate(),
    segmentEndDate: taperEndDate.toScheduleDate()
  }
}

export function getFormattedListForCopyPaste(schedule: Schedule): string {
  const { segments, languageKey } = schedule;
  const selectedLanguage = getLanguageFromKey(languageKey);

  // Filter out placeholder segments
  const validSegments = segments.filter(segment => !isSegmentPlaceholder(segment));

  // Format each segment
  const formattedSegments = validSegments.map((segment, index) => {
    const { segmentStartDate, segmentEndDate } = calculateSegmentStartAndEndDates(schedule, index);

    return formatSegmentText({
      segment,
      segmentStartDate,
      segmentEndDate,
      index,
      selectedLanguage
    });
  });

  // Combine formatted segments into a single string
  return formattedSegments.join('\n');
}
