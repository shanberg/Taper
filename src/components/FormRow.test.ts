import { describe, test, expect, vi, beforeEach } from 'vitest';
import { render, fireEvent } from '@testing-library/svelte';
import '@testing-library/jest-dom';
import FormRow from './FormRow.svelte'; // Adjust the import path as necessary

// Mock the utility functions
vi.mock('../utils', () => ({
	isRowInvalid: vi.fn(),
	isRowPlaceholder: vi.fn(),
  }));

import { isRowInvalid, isRowPlaceholder } from '../utils';

describe('FormRow', () => {
	let tableData: UITableData;
	let row: Row;
	let index: number;

	beforeEach(() => {
		tableData = [
			{ dose: 1, daysForDose: 2 },
			{ dose: 3, daysForDose: 4 }
		];
		row = { dose: 1, daysForDose: 2 };
		index = 0;
		isRowInvalid.mockReturnValue(false);
		isRowPlaceholder.mockReturnValue(false);
	});

	test('renders correctly', () => {
		const { container } = render(FormRow, { tableData, row, index });
		expect(container).toMatchSnapshot();
	});

	test('handles dose change', async () => {
		const { getByDisplayValue, component } = render(FormRow, { tableData, row, index });
		const input = getByDisplayValue('1');

		const changeHandler = vi.fn();
		component.$on('change', changeHandler);

		await fireEvent.input(input, { target: { value: '2' } });
		await fireEvent.change(input);

		expect(changeHandler).toHaveBeenCalledWith(
			expect.objectContaining({
				detail: { ...row, dose: 2 }
			})
		);
	});

	test('handles daysForDose change', async () => {
		const { getByDisplayValue, component } = render(FormRow, { tableData, row, index });
		const input = getByDisplayValue('2');

		const changeHandler = vi.fn();
		component.$on('change', changeHandler);

		await fireEvent.input(input, { target: { value: '3' } });
		await fireEvent.change(input);

		expect(changeHandler).toHaveBeenCalledWith(
			expect.objectContaining({
				detail: { ...row, daysForDose: 3 }
			})
		);
	});

	test('handles remove row', async () => {
		const { getByTitle, component } = render(FormRow, { tableData, row, index });
		const button = getByTitle('Remove this step');

		const removeHandler = vi.fn();
		component.$on('removeRow', removeHandler);

		await fireEvent.click(button);

		expect(removeHandler).toHaveBeenCalledWith(
			expect.objectContaining({
				detail: index
			})
		);
	});

	test('disables remove button when row is the only real row', () => {
		tableData = [
			{ dose: 1, daysForDose: 1 },
			{ dose: 1, daysForDose: 1 }
		];
		const { getByTitle } = render(FormRow, { tableData, row, index });
		const button = getByTitle('Remove this step');

		expect(button).toBeDisabled();
	});

	test('applies placeholder class when row is a placeholder', () => {
		isRowPlaceholder.mockReturnValue(true);
		const { container } = render(FormRow, { tableData, row, index });

		expect(container.querySelector('.row')).toHaveClass('placeholder');
	});

	test('applies isInvalid class when row is invalid', () => {
		isRowInvalid.mockReturnValue(true);
		const { container } = render(FormRow, { tableData, row, index });

		expect(container.querySelector('.row')).toHaveClass('isInvalid');
	});
});
