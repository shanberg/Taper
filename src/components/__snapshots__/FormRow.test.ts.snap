// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormRow > renders correctly 1`] = `
<div>
  <tr
    class="row   svelte-4q5etb"
  >
    <td
      class="dose"
    >
      <input
        class="svelte-4q5etb"
        inputmode="decimal"
        min="1"
        step="0.25"
        type="number"
      />
    </td>
     
    <td
      class="days"
    >
      <input
        class="svelte-4q5etb"
        inputmode="decimal"
        min="1"
        type="number"
      />
    </td>
     
    <td
      class="delete"
    >
      <button
        class="remove-btn svelte-4q5etb"
        disabled=""
        title="Remove this step"
      >
        Ã—
      </button>
    </td>
  </tr>
</div>
`;
